### PR TITLE
[rubysrc2cpg] Unit tests replicated from c2cpg and pysrc2cpg

### PR DESCRIPTION
**Pull request by pandurangpatil**
ruby and go AST unit tests
1. File node related unit tests - Added similar unit tests for `gosrc2cpg`
2. Metadata unit tests
3. Namespace unit tests
4. Method unit tests covering 
    - Two types of method definitions a. def ... end b. define_method ... end
    - Method parameters and its traversal
    - Method return nodes and its traversal

### CODE REVIEW

File: joern-cli/frontends/rubysrc2cpg/src/test/scala/io/joern/rubysrc2cpg/passes/ast/MethodOneTests.scala [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    "should contain exactly one method node with correct fields" in {
+      inside(cpg.method.name("foo").l) { case List(x) =>
+        x.name shouldBe "foo"
+        x.isExternal shouldBe false
+        x.fullName shouldBe "Test0.rb::program:foo"
+        x.code should startWith("def foo(a, b)")
+        x.isExternal shouldBe false
+        x.order shouldBe 1
+        // TODO: At present it is returning absolute path
+//        x.filename shouldBe "Test0.rb"
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/rubysrc2cpg/src/test/scala/io/joern/rubysrc2cpg/passes/ast/MethodOneTests.scala:27]: Could check `endsWith`

File: joern-cli/frontends/rubysrc2cpg/src/test/scala/io/joern/rubysrc2cpg/passes/ast/MethodOneTests.scala [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+    // TODO: Need to be fixed
+    "test method parameter nodes" ignore {
+      cpg.method.name("foo").parameter.name.l.size shouldBe 2
+      val parameter1 = cpg.method.fullName("Test0.rb::program:foo").parameter.order(1).head
+      parameter1.name shouldBe "a"
+      parameter1.index shouldBe 1
+      parameter1.typeFullName shouldBe "ANY"
+
+      val parameter2 = cpg.method.fullName("test.py:<module>.func").parameter.order(2).head
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/rubysrc2cpg/src/test/scala/io/joern/rubysrc2cpg/passes/ast/MethodOneTests.scala:87]:
```suggestion
      val parameter2 = cpg.method.fullName("Test0.rb::program:foo").parameter.order(2).head
```

File: joern-cli/frontends/rubysrc2cpg/src/test/scala/io/joern/rubysrc2cpg/passes/ast/MethodOneTests.scala [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+    }
+  }
+
+  "Multiple Return tests" should {
+    val cpg = code("""
+        |def foo(names)
+        |  if names == "Alice"
+        |    return 1
+        |  else
+        |    return 1
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/rubysrc2cpg/src/test/scala/io/joern/rubysrc2cpg/passes/ast/MethodOneTests.scala:122]:
```suggestion
        |    return 2
```

File: joern-cli/frontends/rubysrc2cpg/src/test/scala/io/joern/rubysrc2cpg/passes/ast/MethodOneTests.scala [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+      inside(cpg.method("foo").methodReturn.l) { case List(mainMethodReturn) =>
+        mainMethodReturn.typeFullName shouldBe "ANY"
+      }
+      val astReturns  = cpg.method("foo").ast.isReturn.l
+      val cfgReturns  = cpg.method("foo").methodReturn.cfgPrev.l
+      val travReturns = cpg.method("foo").methodReturn.toReturn.l
+      inside(astReturns) { case List(ret1, ret2) =>
+        ret1.code shouldBe "return 1"
+        ret1.lineNumber shouldBe Option(4)
+        ret2.code shouldBe "return 2;"
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/rubysrc2cpg/src/test/scala/io/joern/rubysrc2cpg/passes/ast/MethodOneTests.scala:138]:
```suggestion
        ret2.code shouldBe "return 2"
```

File: joern-cli/frontends/rubysrc2cpg/src/test/scala/io/joern/rubysrc2cpg/passes/ast/MethodTwoTests.scala [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    "should contain exactly one method node with correct fields" ignore {
+      inside(cpg.method.name("foo").l) { case List(x) =>
+        x.name shouldBe "foo"
+        x.isExternal shouldBe false
+        x.fullName shouldBe "Test0.rb::program:foo"
+        x.code should startWith("def foo(a, b)")
+        x.isExternal shouldBe false
+        x.order shouldBe 1
+        // TODO: At present it is returning absolute path
+        //        x.filename shouldBe "Test0.rb"
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/rubysrc2cpg/src/test/scala/io/joern/rubysrc2cpg/passes/ast/MethodTwoTests.scala:27]: Here again we can check `endsWith`

File: joern-cli/frontends/rubysrc2cpg/src/test/scala/io/joern/rubysrc2cpg/passes/ast/MethodTwoTests.scala [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    // TODO: Need to be fixed
+    "test method parameter nodes" ignore {
+
+      cpg.method.name("foo").parameter.name.l.size shouldBe 2
+      val parameter1 = cpg.method.fullName("Test0.rb::program:foo").parameter.order(1).head
+      parameter1.name shouldBe "a"
+      parameter1.index shouldBe 1
+      parameter1.typeFullName shouldBe "ANY"
+
+      val parameter2 = cpg.method.fullName("test.py:<module>.func").parameter.order(2).head
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/rubysrc2cpg/src/test/scala/io/joern/rubysrc2cpg/passes/ast/MethodTwoTests.scala:92]:
```suggestion
      val parameter2 = cpg.method.fullName("Test0.rb::program:foo").parameter.order(2).head
```

File: joern-cli/frontends/rubysrc2cpg/src/test/scala/io/joern/rubysrc2cpg/passes/ast/NamespaceBlockTest.scala [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+  "should allow traversing from namespace block to type declaration" in {
+    cpg.namespaceBlock
+      .filenameNot(FileTraversal.UNKNOWN)
+      .ast
+      .isTypeDecl
+      .nameNot(NamespaceTraversal.globalNamespaceName)
+      .name
+      .l
+      .sorted shouldBe List("MyClass")
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/rubysrc2cpg/src/test/scala/io/joern/rubysrc2cpg/passes/ast/NamespaceBlockTest.scala:43]: Why call `.sorted` if we expect a singleton?

File: joern-cli/frontends/rubysrc2cpg/src/test/scala/io/joern/rubysrc2cpg/passes/ast/MethodTwoTests.scala (hunk 5 above)
Comment (pandurangpatil) [reply joern-cli/frontends/rubysrc2cpg/src/test/scala/io/joern/rubysrc2cpg/passes/ast/MethodTwoTests.scala:27]: @DavidBakerEffendi  what is expected full file path or relative file path?

File: joern-cli/frontends/rubysrc2cpg/src/test/scala/io/joern/rubysrc2cpg/passes/ast/MethodOneTests.scala (hunk 3 above)
Comment (pandurangpatil) [reply joern-cli/frontends/rubysrc2cpg/src/test/scala/io/joern/rubysrc2cpg/passes/ast/MethodOneTests.scala]: done

**Review by DavidBakerEffendi — APPROVED** [review]

**Pull request merged**